src/passes/Heap2Local.cpp

```diff
@@ -647,6 +659,12 @@ struct Heap2LocalOptimizer {
         }
       }
 
+      void visitRefCast(RefCast* curr) {
+        if (Type::isSubType(allocation->type, curr->type)) {
```

Comment:
```suggestion
        // We only accept a value of the allocation's type (so that we know this
        // cast is removable later).
        if (Type::isSubType(allocation->type, curr->type)) {
```

Reply:
Or, rather, it's not about being removable later, but if we cast to something entirely different then we can't access the values of that thing (we only know the layout of the allocation), but the cast will trap anyhow, so we stop.